lib/hooks.js

```diff
@@ -0,0 +1,161 @@
+  this[hook].push(fn)
+}
+
+function assignLifeCycleHooksToContext (hooks, context) {
+  const contextHooks = {
+    preParsing: null,
+    preValidation: null,
+    preExecution: null,
+    preGatewayExecution: null,
+    onResolution: null
+  }
+  if (hooks.preParsing.length > 0) contextHooks.preParsing = hooks.preParsing.slice()
+  if (hooks.preValidation.length > 0) contextHooks.preValidation = hooks.preValidation.slice()
+  if (hooks.preExecution.length > 0) contextHooks.preExecution = hooks.preExecution.slice()
+  if (hooks.preGatewayExecution.length > 0) contextHooks.preGatewayExecution = hooks.preGatewayExecution.slice()
+  if (hooks.onResolution.length > 0) contextHooks.onResolution = hooks.onResolution.slice()
+  return Object.assign(contextHooks, context)
+}
+
+async function hookRunner (functions, runner, request, cb) {
+  let i = 0
+
+  async function next (err) {
+    // If finished, call the handler callback
+    if (err || i === functions.length) {
+      cb(err, request)
+      return
+    }
+
+    let result
+    try {
+      result = runner(functions[i++], request)
+    } catch (error) {
+      next(error)
+      return
+    }
+    if (result && typeof result.then === 'function') {
+      await result.then(handleResolve, handleReject)
+    }
+  }
+
+  async function handleResolve () {
+    await next()
+  }
+
+  function handleReject (err) {
+    cb(err, request)
+  }
+
+  await next()
+}
+
+async function preExecutionHookRunner (functions, request, executionResult, cb) {
+  let i = 0
+
+  async function next (err, newDocument, newErrors) {
+    // If finished, call the handler callback
+    if (err) {
+      cb(err, request, executionResult)
+      return
+    }
+
+    if (newDocument !== undefined) {
+      request.document = newDocument
+      request.modifiedQuery = true
+    }
+
+    if (newErrors !== undefined) {
+      if (executionResult.errors !== undefined) {
+        executionResult.errors = executionResult.errors.concat(newErrors)
+      } else {
+        executionResult.errors = newErrors
+      }
+    }
+
+    if (i === functions.length) {
+      cb(null, request, executionResult)
+      return
+    }
+
+    let result
+    try {
+      result = functions[i++](request.schema, request.document, request.context)
+    } catch (error) {
+      await next(error)
+      return
+    }
+    if (result && typeof result.then === 'function') {
+      await result.then(handleResolve, handleReject)
+    } else {
+      await handleResolve(result)
+    }
+  }
+
+  async function handleResolve (result) {
+    let newDocument
+    let newErrors
+    if (result) {
+      if (typeof result.document !== 'undefined') {
+        newDocument = result.document
+      }
+      if (typeof result.errors !== 'undefined') {
+        newErrors = result.errors
+      }
+    }
+    await next(null, newDocument, newErrors)
+  }
+
+  function handleReject (err) {
+    cb(err, request, executionResult)
+  }
+
+  await next()
```

Comment:
Could you please avoid the recursion and use a loop instead? We are just using promises here, there is no need to have a callback.

Reply:
yep, no problem